### PR TITLE
Attempt to mark EqualityContract property accessor as not auto-implemented

### PR DESCRIPTION
**Pull request by Youssef1313**
No idea if this will fail something. But at a first glance it looks like it shouldn't be an auto implemented property accessor. The property itself isn't auto property, why would the accessor be?

The motivation for the change here is to make `isAutoPropertyAccessor` always matching `property.IsAutoProperty` so we can actually be able to remove the parameter entirely.

This is one step towards making #57915 pass, which will make the life much easier for semi auto props.

@AlekseyTs for review.

### CODE REVIEW

**Comment by Youssef1313** [comment]
I don't know if the `Source-Build (Managed)` can be related to my change. I didn't restart the build in case the failure is unrelated and someone will want to look into the build artifacts.

```
  /__w/1/s/artifacts/source-build/self/package-cache/microsoft.net.compilers.toolset/4.1.0-1.21553.8/tasks/netcoreapp3.1/Microsoft.CSharp.Core.targets(75,5): error MSB3883: Unexpected exception:  [/__w/1/s/artifacts/source-build/self/src/src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj]
##[error]artifacts/source-build/self/package-cache/microsoft.net.compilers.toolset/4.1.0-1.21553.8/tasks/netcoreapp3.1/Microsoft.CSharp.Core.targets(75,5): error MSB3883: (NETCORE_ENGINEERING_TELEMETRY=Build) Unexpected exception: 
  /__w/1/s/artifacts/source-build/self/package-cache/microsoft.net.compilers.toolset/4.1.0-1.21553.8/tasks/netcoreapp3.1/Microsoft.CSharp.Core.targets(75,5): error : One or more errors occurred. (ReleaseMutex failed. WaitOne Id: 13 Release Id: 13) [/__w/1/s/artifacts/source-build/self/src/src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj]
##[error]artifacts/source-build/self/package-cache/microsoft.net.compilers.toolset/4.1.0-1.21553.8/tasks/netcoreapp3.1/Microsoft.CSharp.Core.targets(75,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) One or more errors occurred. (ReleaseMutex failed. WaitOne Id: 13 Release Id: 13)
    CSharpSyntaxGenerator -> /__w/1/s/artifacts/source-build/self/src/artifacts/bin/CSharpSyntaxGenerator/Release/netstandard2.0/CSharpSyntaxGenerator.dll
    CSharpSyntaxGenerator -> /__w/1/s/artifacts/source-build/self/src/artifacts/bin/CSharpSyntaxGenerator/Release/net6.0/CSharpSyntaxGenerator.dll
    Microsoft.Build.Tasks.CodeAnalysis -> /__w/1/s/artifacts/source-build/self/src/artifacts/bin/Microsoft.Build.Tasks.CodeAnalysis/Release/net472/Microsoft.Build.Tasks.CodeAnalysis.dll
    Microsoft.Build.Tasks.CodeAnalysis -> /__w/1/s/artifacts/source-build/self/src/artifacts/bin/Microsoft.Build.Tasks.CodeAnalysis/Release/net6.0/Microsoft.Build.Tasks.CodeAnalysis.dll
    Microsoft.CodeAnalysis -> /__w/1/s/artifacts/source-build/self/src/artifacts/bin/Microsoft.CodeAnalysis/Release/netcoreapp3.1/Microsoft.CodeAnalysis.dll
    Microsoft.Build.Tasks.CodeAnalysis -> /__w/1/s/artifacts/source-build/self/src/artifacts/bin/Microsoft.Build.Tasks.CodeAnalysis/Release/net6.0/publish/
  
  Build FAILED.
  
  /__w/1/s/artifacts/source-build/self/package-cache/microsoft.net.compilers.toolset/4.1.0-1.21553.8/tasks/netcoreapp3.1/Microsoft.CSharp.Core.targets(75,5): error MSB3883: Unexpected exception:  [/__w/1/s/artifacts/source-build/self/src/src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj]
  /__w/1/s/artifacts/source-build/self/package-cache/microsoft.net.compilers.toolset/4.1.0-1.21553.8/tasks/netcoreapp3.1/Microsoft.CSharp.Core.targets(75,5): error : One or more errors occurred. (ReleaseMutex failed. WaitOne Id: 13 Release Id: 13) [/__w/1/s/artifacts/source-build/self/src/src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj]
      0 Warning(s)
      2 Error(s)
  
  Time Elapsed 00:01:19.30
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/SourceBuild/SourceBuildArcadeBuild.targets(167,5): error MSB3073: The command "/__w/1/s/.dotnet/dotnet msbuild /m /nologo /clp:Summary /v:minimal /nr:false /p:TreatWarningsAsErrors=false /p:ContinuousIntegrationBuild=true /__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/Build.proj /bl:"/__w/1/s/artifacts/log/Release/Build.binlog" /p:Configuration=Release /p:Projects=/__w/1/s/Compilers.sln /p:RepoRoot=/__w/1/s/ /p:Restore=true /p:Build=true /p:Rebuild=false /p:Test=false /p:Pack=true /p:Publish=true /p:RunAnalyzersDuringBuild=false /p:BootstrapBuildPath= /p:ContinuousIntegrationBuild=true /p:TreatWarningsAsErrors=true /p:TestRuntimeAdditionalArguments= /p:ArcadeBuildFromSource=true /p:ROSLYNUSEHARDLINKS=true /p:SourceBuildNonPortable= -logger:/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/netcoreapp3.1/Microsoft.DotNet.ArcadeLogging.dll  /p:Projects="/__w/1/s/artifacts/source-build/self/src/Roslyn.sln" /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir= /p:EnableSourceControlManagerQueries=false /p:EnableSourceLink=false /p:DeterministicSourcePaths=false" exited with code 1. [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/Build.proj]
##[error].packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/SourceBuild/SourceBuildArcadeBuild.targets(167,5): error MSB3073: The command "/__w/1/s/.dotnet/dotnet msbuild /m /nologo /clp:Summary /v:minimal /nr:false /p:TreatWarningsAsErrors=false /p:ContinuousIntegrationBuild=true /__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/Build.proj /bl:"/__w/1/s/artifacts/log/Release/Build.binlog" /p:Configuration=Release /p:Projects=/__w/1/s/Compilers.sln /p:RepoRoot=/__w/1/s/ /p:Restore=true /p:Build=true /p:Rebuild=false /p:Test=false /p:Pack=true /p:Publish=true /p:RunAnalyzersDuringBuild=false /p:BootstrapBuildPath= /p:ContinuousIntegrationBuild=true /p:TreatWarningsAsErrors=true /p:TestRuntimeAdditionalArguments= /p:ArcadeBuildFromSource=true /p:ROSLYNUSEHARDLINKS=true /p:SourceBuildNonPortable= -logger:/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/netcoreapp3.1/Microsoft.DotNet.ArcadeLogging.dll  /p:Projects="/__w/1/s/artifacts/source-build/self/src/Roslyn.sln" /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir= /p:EnableSourceControlManagerQueries=false /p:EnableSourceLink=false /p:DeterministicSourcePaths=false" exited with code 1.

Build FAILED.

/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/SourceBuild/SourceBuildArcadeBuild.targets(167,5): error MSB3073: The command "/__w/1/s/.dotnet/dotnet msbuild /m /nologo /clp:Summary /v:minimal /nr:false /p:TreatWarningsAsErrors=false /p:ContinuousIntegrationBuild=true /__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/Build.proj /bl:"/__w/1/s/artifacts/log/Release/Build.binlog" /p:Configuration=Release /p:Projects=/__w/1/s/Compilers.sln /p:RepoRoot=/__w/1/s/ /p:Restore=true /p:Build=true /p:Rebuild=false /p:Test=false /p:Pack=true /p:Publish=true /p:RunAnalyzersDuringBuild=false /p:BootstrapBuildPath= /p:ContinuousIntegrationBuild=true /p:TreatWarningsAsErrors=true /p:TestRuntimeAdditionalArguments= /p:ArcadeBuildFromSource=true /p:ROSLYNUSEHARDLINKS=true /p:SourceBuildNonPortable= -logger:/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/netcoreapp3.1/Microsoft.DotNet.ArcadeLogging.dll  /p:Projects="/__w/1/s/artifacts/source-build/self/src/Roslyn.sln" /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir= /p:EnableSourceControlManagerQueries=false /p:EnableSourceLink=false /p:DeterministicSourcePaths=false" exited with code 1. [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.21566.10/tools/Build.proj]
    0 Warning(s)
    1 Error(s)

```

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs:137]: >isAutoPropertyAccessor [](http://example.com/codeflow?start=23&length=22)

It looks like we are going to stop synthesizing ```CompilerGeneratedAttribute``` for this method. Should be easy to mitigate, I think.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs:137]: @AlekseyTs I'm feeling the attribute should be on the whole property, not just the accessor. What do you think?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:1836]: While the attribute exists in the test, the following lines produce an empty `ImmutableArray<AttributeData>`:

```csharp
var member = verifier.Compilation.GetTypeByMetadataName("C").GetMembers("get_EqualityContract").Single();
var attributes = member.GetAttributes();
```

I have no idea why.

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:1836]: Found that other tests operate on metadata symbols when working with `GetAttributes`. I updated the test, though I'm interested in understanding more on that if someone have the time :)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs:150]: >AddSynthesizedAttribute [](http://example.com/codeflow?start=16&length=23)

consider adding ```Debug.Assert(WellKnownMembers.IsSynthesizedAttributeOptional(WellKnownMember.System_Runtime_CompilerServices_CompilerGeneratedAttribute__ctor));```

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:1817]: >VerifyTypeIL [](http://example.com/codeflow?start=21&length=12)

It feels like the use of this method is an overkill. Especially if our goal is to simply check presence of ```CompilerGeneratedAttribute```, ```symbolValidator``` is good enough.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:1836]: > though I'm interested in understanding more on that if someone have the time

For symbols from source, ```GetAttributes``` never includes attributes that compiler is going to add itself, its result reflects what is in source. The ```symbolValidator``` operates on symbols that we load from the emitted binary for validation purposes. ```GetAttributes``` for those symbols reflects all attributes present in metadata, i.e. all attributes that were actually emitted.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5).

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs:150]: I don't understand the value of this assertion tbh, but I'm adding it anyway.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEqualityContractProperty.cs:150]: >I don't understand the value of this assertion tbh

It is Ok to not report an error if a synthesized attribute is optional and "bad" in one way or another.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 7), assuming CI is passing.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For a second review.

**Review by 333fred — APPROVED** [review]
LGTM (commit 7)

**Pull request merged**
